File: js-template/leetcode/editor/cn/spiral-matrix.js
/*
 * @lc app=leetcode.cn id=54 lang=javascript
 * @lcpr version=30404
 *
 * [54] 螺旋矩阵
 */


import {ListNode} from "../common/listNode.js";
import {TreeNode} from "../common/treeNode.js";

// @lc code=start
/**
 * @param {number[][]} matrix
 * @return {number[]}
 */
var spiralOrder = function(matrix) {
    var direction = ["right", "down", "left", "up"];
    var direc_index = 0;
    var row = matrix.length;
    var col = matrix[0].length;
    var next = [0, 0];
    var result = [];

    const testNext = function(m, n, direct, matrix) {
        let testm = m;
        let testn = n;

        if (direct === "right") {
            testn++;            
        } else if (direct === "down") {
            testm++;
        } else if (direct === "left") {
            testn--;
        } else if (direct === "up") {
            testm--;
        }
        if (testm < row && testn < col && matrix[testm][testn] != null)
            return [testm, testn];
        else
            return undefined;
    }

    while (next[0] < row && next[1] < col && matrix[next[0]][next[1]] != null) {
        let m = next[0];
        let n = next[1];
        result.push(matrix[m][n]);
        matrix[m][n] = undefined;
        
        let v = testNext(m, n, direction[direc_index], matrix);
        if (v) {
            next[0] = v[0];
            next[1] = v[1];
        } else {
            direc_index = (direc_index + 1) % direction.length;
            v = testNext(m, n, direction[direc_index], matrix);
            next[0] = v ? v[0] : undefined;
            next[1] = v ? v[1] : undefined;
        }
    }
    return result;
};
// @lc code=end

// your test code here

spiralOrder([[2,5],[8,4],[0,-1]])



/*
// @lcpr case=start
// [[1,2,3],[4,5,6],[7,8,9]]\n
// @lcpr case=end

// @lcpr case=start
// [[1,2,3,4],[5,6,7,8],[9,10,11,12]]\n
// @lcpr case=end

 */

File: cpp-template/leetcode/editor/cn/spiral-matrix.cpp
/*
 * @lc app=leetcode.cn id=54 lang=cpp
 * @lcpr version=30404
 *
 * [54] 螺旋矩阵
 */

#include <iostream>
#include <vector>
#include <string>
#include "../common/ListNode.cpp"
#include "../common/TreeNode.cpp"

using namespace std;

// @lc code=start
class Solution {
public:
    vector<int> spiralOrder(vector<vector<int>>& matrix) {
        vector<int> result;
        int m = matrix.size();
        int n = matrix[0].size();
        int maxCount = m * n;
        int i = 0, j = 0, border = 0, count = 1;
        int direction = 0;

        result.push_back(matrix[i][j]);

        while (count < maxCount)
        {
            int test_i = i;
            int test_j = j;
            direction %= 4;
            if (direction == 0) {
                ++test_j;
            } else if (direction == 1) {
                ++test_i;
            } else if (direction == 2) {
                --test_j;
            } else if (direction == 3) {
                --test_i;
            }
            if (test_i == border && test_j == border) {
                ++border;
                ++direction;
                ++j;
                result.push_back(matrix[i][j]);
                ++count;
            } else if (border > test_i || test_i >= m - border || border > test_j || test_j >= n - border) {
                ++direction;
            } else {
                i = test_i;
                j = test_j;
                result.push_back(matrix[i][j]);
                ++count;
            }
        }
       
        return result; 
    }
};
// @lc code=end

int main() {
    Solution solution;
    // your test code here
    vector<vector<int>> matrix = {{1,2,3},{4,5,6},{7,8,9}};
    solution.spiralOrder(matrix);
}



/*
// @lcpr case=start
// [[1,2,3],[4,5,6],[7,8,9]]\n
// @lcpr case=end

// @lcpr case=start
// [[1,2,3,4],[5,6,7,8],[9,10,11,12]]\n
// @lcpr case=end

 */

